### PR TITLE
Simplify control flow in `_Traits_find_MEOW_not_of`

### PR DESCRIPTION
**Pull request by StephanTLavavej**
This simplifies `_Traits_find_first_not_of` and `_Traits_find_last_not_of`, following the patterns in #4744 and #4934, which will make them easier to understand and easier to add vectorized codepaths. These are lossless transformations that don't affect behavior.

* Replace unnecessarily confusing `break`s with immediate `return`s.
  + Except in `_Traits_find_last_of`, which is being changed by #4934.
* Add a `return` instead of relying on fall-through.
  + This will simplify the following transformation.
* Flip `!_Matches._Mark` tests.
  + Both sides unconditionally return, allowing us to flip them. (One case is a forever loop where the only ways out are returns.)
* Transform "call self with `_Special=false`" to fall-through.
* Extract `_Elem`.
* Extract `_Hay_start`, `_Hay_end`.
* Fuse `bool _Special` into its usage.

### CODE REVIEW

**Review by AlexGuteniev — CHANGES_REQUESTED** [review]
I was going to do most of this in `not` vectorization PR, but maybe it is indeed cleaner not to mix.

As we no longer have tail calls here, remove `bool _Special = _Is_implementation_handled_char_traits<_Traits>` template parameter, as @frederick-vs-ja suggested and attempted.

**Comment by StephanTLavavej** [comment]
> I was going to do most of this in not vectorization PR, but maybe it is indeed cleaner not to mix.

Yeah, it's difficult for me to follow two major changes combined, and I have a pretty high tolerance for review complexity.

> As we no longer have tail calls here, remove `bool _Special = _Is_implementation_handled_char_traits<_Traits>` template parameter, as @frederick-vs-ja suggested and attempted.

I agree that this should be done after #4934 and this PR land.

File: stl/inc/__msvc_string_view.hpp [hunk 1]
```diff
@@ -635,7 +635,7 @@ constexpr size_t _Traits_rfind(_In_reads_(_Hay_size) const _Traits_ptr_t<_Traits
             }
 
             if (_Match_try == _Haystack) {
-                break; // at beginning, no more chance for match
+                return static_cast<size_t>(-1); // at beginning, no more chance for match
```
Comment (CaseyCarter) [inline stl/inc/__msvc_string_view.hpp]: It seems odd to me that we're happy to early-return in this function, but we don't do so on line 631.

File: stl/inc/__msvc_string_view.hpp [hunk 2]
```diff
@@ -654,7 +654,7 @@ constexpr size_t _Traits_rfind_ch(_In_reads_(_Hay_size) const _Traits_ptr_t<_Tra
             }
```
Comment (CaseyCarter) [inline stl/inc/__msvc_string_view.hpp]: Ditto "early return on 650 could save us an indent".

File: stl/inc/__msvc_string_view.hpp [hunk 3]
```diff
@@ -723,7 +723,7 @@ constexpr size_t _Traits_find_first_of(_In_reads_(_Hay_size) const _Traits_ptr_t
         const auto _Hay_start = _Haystack + _Start_at;
```
Comment (CaseyCarter) [inline stl/inc/__msvc_string_view.hpp]: Ditto "early return on 722 could save us an indent".

File: stl/inc/__msvc_string_view.hpp [hunk 4]
```diff
@@ -814,32 +814,33 @@ constexpr size_t _Traits_find_last_of(_In_reads_(_Hay_size) const _Traits_ptr_t<
     return static_cast<size_t>(-1); // no match
 }
 
-template <class _Traits, bool _Special = _Is_implementation_handled_char_traits<_Traits>>
+template <class _Traits>
 constexpr size_t _Traits_find_first_not_of(_In_reads_(_Hay_size) const _Traits_ptr_t<_Traits> _Haystack,
     const size_t _Hay_size, const size_t _Start_at, _In_reads_(_Needle_size) const _Traits_ptr_t<_Traits> _Needle,
     const size_t _Needle_size) noexcept {
     // in [_Haystack, _Haystack + _Hay_size), look for none of [_Needle, _Needle + _Needle_size), at/after _Start_at
     if (_Start_at < _Hay_size) { // room for match, look for it
```
Comment (CaseyCarter) [inline stl/inc/__msvc_string_view.hpp]: Ditto "Consider early return".

File: stl/inc/__msvc_string_view.hpp [hunk 5]
```diff
@@ -863,37 +864,39 @@ constexpr size_t _Traits_find_not_ch(_In_reads_(_Hay_size) const _Traits_ptr_t<_
     return static_cast<size_t>(-1); // no match
 }
 
-template <class _Traits, bool _Special = _Is_implementation_handled_char_traits<_Traits>>
+template <class _Traits>
 constexpr size_t _Traits_find_last_not_of(_In_reads_(_Hay_size) const _Traits_ptr_t<_Traits> _Haystack,
     const size_t _Hay_size, const size_t _Start_at, _In_reads_(_Needle_size) const _Traits_ptr_t<_Traits> _Needle,
     const size_t _Needle_size) noexcept {
     // in [_Haystack, _Haystack + _Hay_size), look for none of [_Needle, _Needle + _Needle_size), before _Start_at
     if (_Hay_size != 0) { // worth searching, do it
```
Comment (CaseyCarter) [inline stl/inc/__msvc_string_view.hpp]: And one last recommendation for early-return.

**Review by CaseyCarter — APPROVED** [review]
This is such a nice simplification, I thought I'd push another commit to do some early-return outdenting.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
:heart_eyes_cat: Looks purr to me!

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by CaseyCarter** [comment]
A merge conflict? Couldn't be me.

**Comment by StephanTLavavej** [comment]
Remember, I named you Casey "Conflict" Carter via office post-it note almost a decade ago! :dart: :100: :crystal_ball:

**Pull request merged**